src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py

```diff
@@ -12,3 +40,198 @@ class InTotoV1Statement(TypedDict):
+
+
+def validate_intoto_statement(payload: dict[str, JsonType]) -> TypeGuard[InTotoV1Statement]:
+    """Validate the statement of an in-toto attestation.
+
+    Specification: https://github.com/in-toto/attestation/tree/main/spec/v1/statement.md.
+
+    Parameters
+    ----------
+    payload : dict[str, JsonType]
+        The JSON statement after being base64-decoded.
+
+    Returns
+    -------
+    TypeGuard[InTotoStatement]
+        ``True`` if the attestation statement is valid, in which case its type is narrowed to an
+        ``InTotoStatement``; ``False`` otherwise.
+
+    Raises
+    ------
+    ValidateInTotoPayloadError
+        When the payload does not follow the expected schema.
+    """
+    type_ = payload.get("_type")
+    if type_ is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute '_type' of the in-toto statement is missing.",
+        )
+    if not isinstance(type_, str) or not type_ == "https://in-toto.io/Statement/v1":
+        raise ValidateInTotoPayloadError(
+            "The value of attribute '_type' in the in-toto statement must be: 'https://in-toto.io/Statement/v1'",
+        )
+
+    subjects_payload = payload.get("subject")
+    if subjects_payload is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute 'subject' of the in-toto statement is missing.",
+        )
+    if not isinstance(subjects_payload, list):
+        raise ValidateInTotoPayloadError(
+            "The value of attribute 'subject' in the in-toto statement is invalid: expecting a list.",
+        )
+
+    for subject_json in subjects_payload:
+        validate_intoto_subject(subject_json)
+
+    predicate_type = payload.get("predicateType")
+    if predicate_type is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute 'predicateType' of the in-toto statement is missing.",
+        )
+
+    if not isinstance(predicate_type, str):
+        raise ValidateInTotoPayloadError(
+            "The value of attribute 'predicateType' in the in-toto statement is invalid: expecting a string."
+        )
+
+    predicate = payload.get("predicate")
+    if predicate is not None and not isinstance(predicate, dict):
+        raise ValidateInTotoPayloadError(
+            "The value attribute 'predicate' in the in-toto statement is invalid: expecting an object.",
+        )
+
+    return True
+
+
+def validate_intoto_subject(subject: JsonType) -> TypeGuard[InTotoV1ResourceDescriptor]:
+    """Validate a single subject in the in-toto statement.
+
+    See specification: https://github.com/in-toto/attestation/blob/main/spec/v1/resource_descriptor.md
+
+    Parameters
+    ----------
+    subject : JsonType
+        The JSON element representing a single subject.
+
+    Returns
+    -------
+    TypeGuard[InTotoSubject]
+        ``True`` if the subject element is valid, in which case its type is narrowed to an
+        ``InTotoSubject``; ``False`` otherwise.
```

Comment:
```suggestion
    Returns
    -------
    TypeGuard[InTotoV1ResourceDescriptor]
        ``True`` if the subject element is valid, in which case its type is narrowed to an
        ``InTotoV1ResourceDescriptor``; ``False`` otherwise.
```